### PR TITLE
fix: avoid overflow in Scan::derive_stats for full-range UInt64 stats

### LINKED ISSUE #19555: bug: planner panic on UInt64 full-range column stats: `attempt to add with overflow` in `Scan::derive_stats`
## Summary

Databend can panic during planning when scanning a `UInt64` column whose statistics span the full unsigned range `[0,
18446744073709551615]`.

The panic happens in `Scan::derive_stats` when NDV is reduced using:

```rust
(Datum::UInt(m), Datum::UInt(n)) if m >= n => ndv.reduce((m - n + 1) as _),

If m = u64::MAX and n = 0, then m - n + 1 overflows.
```
## Environment

- Branch: main

## Minimal Reproduction
```
DROP TABLE IF EXISTS t;

CREATE TABLE t AS
SELECT d::UInt64 AS d
FROM (VALUES
    (0),
    (18446744073709551615)
) AS v(d);

SELECT d FROM t;
```
## Actual Result

The query fails and the server logs a panic similar to:

panicked at src/query/sql/src/planner/plans/scan.rs:285:78:
attempt to add with overflow

I also reproduced the same failure with:
```
SELECT * FROM t;
SELECT count(*) FROM t;
EXPLAIN SELECT d FROM t;
```
## Expected Result

Databend should not panic during planning.

The query should either:

- succeed normally, or
- return a regular query error if the case is unsupported

but it should not crash due to integer overflow.

## Suspected Root Cause

In src/query/sql/src/planner/plans/scan.rs:

(Datum::UInt(m), Datum::UInt(n)) if m >= n => ndv.reduce((m - n + 1) as _),

For UInt64 stats with:

- min = 0
- max = 18446744073709551615

the expression m - n + 1 becomes u64::MAX + 1, which overflows.

## Suggested Fix

Use checked or saturating arithmetic when reducing NDV for unsigned integer ranges, for example:

let span = m.saturating_sub(*n).saturating_add(1);

or equivalent checked logic that clamps safely.

## Notes

This is not limited to the SQL above. Any table/imported data source that produces UInt64 column stats over the full domain can
trigger the same planner panic.

### PR DESCRIPTION
**Pull request by sundy-li**
I hereby agree to the terms of the CLA available at: https://docs.databend.com/dev/policies/cla/

## Summary

- fix the inclusive integer-span calculation in `Scan::derive_stats` so full-range `UInt64` stats do not overflow during NDV reduction
- reuse the same overflow-safe span logic for signed integer stats, which also avoids the existing `i64::MAX` saturation edge case
- add a planner regression test that reproduces the issue through `CollectStatisticsOptimizer` with `min = 0`, `max = u64::MAX`, and `ndv = 2`

## Tests

- [x] Unit Test
- [ ] Logic Test
- [ ] Benchmark Test
- [ ] No Test - _Explain why_

Verification:
- `cargo fmt --all --check`
- `cargo test -p databend-common-sql --test it planner::fixture::tests::`
- `cargo clippy -p databend-common-sql --test it -- -D warnings`

## Type of change

- [x] Bug Fix (non-breaking change which fixes an issue)
- [ ] New Feature (non-breaking change which adds functionality)
- [ ] Breaking Change (fix or feature that could cause existing functionality not to work as expected)
- [ ] Documentation Update
- [ ] Refactoring
- [ ] Performance Improvement
- [ ] Other (please describe):

Fixes #19555

- - -
This change is [Reviewable](https://reviewable.io/reviews/databendlabs/databend/19591)

### CODE REVIEW

File: src/query/sql/src/planner/plans/scan.rs [hunk 1]
```diff
@@ -283,10 +295,7 @@ impl Operator for Scan {
 
                 // Alter ndv based on min and max if the datum is uint or int.
                 let ndv = match (&max, &min) {
-                    (Datum::UInt(m), Datum::UInt(n)) if m >= n => ndv.reduce((m - n + 1) as _),
-                    (Datum::Int(m), Datum::Int(n)) if m >= n => {
-                        ndv.reduce(m.saturating_add(1).saturating_sub(*n) as _)
-                    }
+                    _ if let Some(span) = inclusive_integer_span(&max, &min) => ndv.reduce(span),
```
Comment (sundy-li) [inline src/query/sql/src/planner/plans/scan.rs:299]: This stops the overflow, but it also makes full-range `UInt64` stats reachable for histogram synthesis. `HistogramBuilder::from_ndv` casts the bounds to `f64` before bucketization, so literals near the top of the domain (for example `u64::MAX - 2`) round to the same value as `u64::MAX`. When `ndv > 2`, `compute_histogram_comparison` can then treat `d > u64::MAX - 2` as `d > max` and estimate zero rows. The new regression test uses `ndv = 2`, so it never exercises this path. Please either skip histogram construction for ranges that exceed exact `f64` integer precision or keep the bounds in an integer representation here.

**Comment by sundy-li** [comment]
Blocking issue: `inclusive_integer_span` fixes the overflow, but it also makes full-range `UInt64` stats reach histogram synthesis. That path still downcasts bounds and literals to `f64`, so tail values like `u64::MAX - 2`, `u64::MAX - 1`, and `u64::MAX` collapse together. With `ndv > 2`, predicates such as `d > u64::MAX - 2` can then be estimated as zero rows even though matching values can exist. The new regression test uses `ndv = 2`, so it never covers this path. Please either skip histogram construction for integer ranges beyond exact `f64` precision or keep integer bounds through histogram/selectivity instead of casting to float.

File: src/query/sql/src/planner/plans/scan.rs (hunk 1 above)
Comment (sundy-li) [reply src/query/sql/src/planner/plans/scan.rs:299]: Addressed in `ea366936fc` by taking the smaller fix path: synthesized integer histograms are now skipped whenever the bounds exceed exact `f64` integer precision. The regression coverage now uses `ndv = 4` on `0..=u64::MAX`, and there is a companion test that verifies exact-range `UInt64` stats still synthesize histograms.

[the remark this answers is not in the document]
Comment (sundy-li) [reply src/query/sql/src/planner/plans/scan.rs:299]: Fixed in `ea366936fc`. `Scan::derive_stats` now skips synthesized integer histograms when the min/max bounds fall outside exact `f64` integer precision, so full-range `UInt64` stats no longer reach the lossy histogram path. Added a regression test with `ndv = 4` for `0..=u64::MAX` and a positive test that safe `UInt64` bounds still build histograms.

**Comment by sundy-li** [comment]
Skipping histogram synthesis here avoids the immediate overflow and the unsafe histogram path, but filtered scans on the same full-range integer stats are still misestimated. Once `histogram` is `None`, `SelectivityEstimator::compute_ndv_comparison()` falls back to `Datum::as_double()` for `min`, `max`, and the literal. For example, `u64::MAX - 2`, `u64::MAX - 1`, and `u64::MAX` all collapse to the same `f64`, so a predicate like `d > u64::MAX - 2` is still estimated as 0 rows even though `u64::MAX` can match. The new fixture only covers the no-filter path, so it doesn't catch this. Please either make the non-histogram integer comparison path precision-safe too, or disable NDV range selectivity when integer bounds are outside exact `f64` precision.

**Comment by sundy-li** [comment]
Blocking issue in src/query/sql/src/planner/optimizer/ir/stats/selectivity.rs:399-405: once src/query/sql/src/planner/plans/scan.rs:323-326 skips histogram synthesis for large integer bounds, filtered scans still fall back to compute_ndv_comparison(), which converts min, max, and the literal to f64. For UInt64(0..=u64::MAX), u64::MAX - 2 and u64::MAX collapse to the same f64, so a > u64::MAX - 2 returns zero selectivity even though u64::MAX matches. I reproduced this on commit ea366936fc with cargo test -p databend-common-sql test_comparison_with_full_range_uint64_keeps_positive_selectivity -- --nocapture, which fails with: expected positive selectivity, got estimated_rows=0. Please make the non-histogram integer comparison path precision-safe too, or disable NDV/range selectivity when bounds exceed exact f64 precision.

**Comment by sundy-li** [comment]
Blocking issue: skipping synthesized histograms for out-of-range integers is only a partial fix here.

After `scan.rs` returns `None` at the new `has_exact_integer_histogram_bounds()` guard, prewhere selectivity falls back to `compute_ndv_comparison()` in `src/query/sql/src/planner/optimizer/ir/stats/selectivity.rs:399-455`, which still converts `min`, `max`, and the predicate literal through `Datum::as_double()`. For full-range `UInt64` stats, values near the top of the domain collapse to the same `f64` (`float(u64::MAX - 2) == float(u64::MAX)`), so a predicate like `d > u64::MAX - 2` is treated as `d > max` and estimated as zero rows even though matching values can exist.

The new tests only cover the no-predicate planner path and the histogram gate, so this remaining path stays untested. Please either keep the NDV comparison math in integer space for unsafe integer bounds or bail out of the NDV-based comparison when the bounds/literal are outside exact `f64` integer precision.

File: src/query/sql/src/planner/plans/scan.rs [hunk 2]
```diff
@@ -297,6 +321,9 @@ impl Operator for Scan {
                     histogram.clone()
                 } else {
                     num_rows.and_then(|num_rows| {
+                        if !has_exact_integer_histogram_bounds(&min, &max) {
```
Comment (sundy-li) [inline src/query/sql/src/planner/plans/scan.rs]: Blocking: returning `None` here avoids the synthesized-histogram overflow, but filtered scans still fall back to `SelectivityEstimator::compute_ndv_comparison()`, which converts `min`, `max`, and the predicate literal through `Datum::as_double()`. For `UInt64(0..=u64::MAX)`, values near the top of the domain collapse to the same `f64`, so a predicate like `d > u64::MAX - 2` is still estimated as zero rows even though `u64::MAX` can match. I reproduced this locally on `ea366936fc` with a focused unit test over `ColumnStat { min: 0, max: u64::MAX, ndv: 4, histogram: None }`. This needs either a precision-safe integer comparison path when histogram synthesis is skipped, or a bailout from NDV/range selectivity for out-of-range integer bounds.

File: src/query/sql/src/planner/plans/scan.rs [hunk 3]
```diff
@@ -175,6 +175,33 @@ impl Scan {
...
+            Some((*max as i128 - *min as i128 + 1) as f64)
+        }
+        _ => None,
+    }
+}
+
+fn has_exact_integer_histogram_bounds(min: &Datum, max: &Datum) -> bool {
+    match (min, max) {
+        (Datum::UInt(min), Datum::UInt(max)) => {
+            *min <= MAX_EXACT_F64_INTEGER && *max <= MAX_EXACT_F64_INTEGER
```
Comment (sundy-li) [inline src/query/sql/src/planner/plans/scan.rs]: `MAX_EXACT_F64_INTEGER` cannot be admitted here with `<=`. `HistogramBuilder::from_ndv()` immediately casts these bounds to `f64`, and `UniformSampleSet::get_upper_bound()` then does `max + 1.0` to model the inclusive integer range. For `max == 1 << 53` (and the signed `+/-` limit), `max + 1.0` rounds back to `max`, so this still synthesizes a lossy histogram on the exact boundary the new guard is trying to exclude. Please tighten this to require `max < 1 << 53` (or otherwise guard on `max + 1` being exactly representable).

**Comment by sundy-li** [comment]
Blocking issue in `src/query/sql/src/planner/plans/scan.rs`:

`has_exact_integer_histogram_bounds()` is still too permissive at the IEEE-754 boundary. It allows `2^53` (and `-2^53..=2^53` for signed values), but `HistogramBuilder::from_ndv()` immediately casts to `f64` and `UniformSampleSet::get_upper_bound()` does `max + 1.0` in float space. At `max == 2^53`, `max + 1.0` rounds back to `2^53`, so a range like `[9007199254740990, 9007199254740992]` still produces duplicated bucket bounds and wrong selectivity estimates instead of the intended safe skip.

Concrete example: for three distinct values `{9007199254740990, 9007199254740991, 9007199254740992}`, the generated histogram makes `< 9007199254740991` estimate `2/3` instead of `1/3`.

Please tighten the guard to exclude that boundary (or move the inclusive-range math out of float space) and add a regression around the edge value.

**Comment by sundy-li** [comment]
Blocking: skipping synthesized histograms in `Scan::derive_stats` avoids the overflow, but filtered scans still fall back to `SelectivityEstimator::compute_ndv_comparison()` with `histogram = None`, and that path downcasts `min`, `max`, and the literal through `Datum::as_double()`.

I verified this on the current PR head with a focused local reproduction over `ColumnStat { min: UInt64(0), max: UInt64(u64::MAX), ndv: 4, histogram: None }`: `a > u64::MAX - 2` is estimated as `0` rows because `u64::MAX - 2` and `u64::MAX` collapse to the same `f64`. That means the PR still leaves incorrect selectivity for the filtered full-range `UInt64` case it is trying to make safe.

This needs either a precision-safe integer comparison path in `compute_ndv_comparison()` or a bailout there when the integer bounds/literal exceed exact `f64` precision.

**Comment by sundy-li** [comment]
Blocking issues:

1. `Scan::derive_stats()` now skips synthesized histograms for wide integer bounds, but that path still falls back to `SelectivityEstimator::compute_ndv_comparison()` when a filter is applied. That code converts `min`, `max`, and the predicate literal through `Datum::as_double()` (`src/query/sql/src/planner/optimizer/ir/stats/selectivity.rs:399-455`). For `UInt64(0..=u64::MAX)`, `float(u64::MAX - 2) == float(u64::MAX)`, so a predicate like `d > u64::MAX - 2` is still estimated as zero rows even though matching values exist. The overflow is gone, but the planner is still unsound for the same full-range stats when histogram synthesis is skipped.

2. `has_exact_integer_histogram_bounds()` still admits the exact `2^53` boundary via `<=` in `src/query/sql/src/planner/plans/scan.rs:195-199`. `HistogramBuilder::from_ndv()` immediately casts those bounds to `f64` (`src/query/sql/src/planner/optimizer/ir/stats/histogram.rs:62-64`) and the float histogram path models inclusive ranges with `max + 1.0`. At `max == 1 << 53`, `max + 1.0 == max`, so the synthesized histogram is already lossy on the boundary this guard is supposed to exclude.

**Comment by sundy-li** [comment]
Blocking: the new histogram gate in `Scan::derive_stats()` fixes the overflow, but filtered scans on the same full-range `UInt64` stats still fall back to `SelectivityEstimator::compute_ndv_comparison()` in `src/query/sql/src/planner/optimizer/ir/stats/selectivity.rs`, which converts `min`, `max`, and the literal through `Datum::as_double()`.

I reproduced this on the current PR head `ea366936fc` with a focused local test over `ColumnStat { min: UInt64(0), max: UInt64(u64::MAX), ndv: 4, histogram: None }`: `a > u64::MAX - 2` is estimated as `0` rows because `u64::MAX - 2` and `u64::MAX` collapse to the same `f64`.

Command used:
`cargo test -p databend-common-sql test_comparison_with_full_range_uint64_keeps_positive_selectivity -- --nocapture`

Failure:
`expected positive selectivity, got estimated_rows=0`

This still leaves incorrect selectivity for the filtered full-range `UInt64` case. Please make the non-histogram integer comparison path precision-safe too, or bail out there when the integer bounds/literal exceed exact `f64` precision.

File: src/query/sql/src/planner/plans/scan.rs (hunk 2 above)
Comment (sundy-li) [inline src/query/sql/src/planner/plans/scan.rs]: Blocking: skipping synthesized histograms here still leaves full-range integer filters on the unsafe `f64` path. Once this returns `None`, `SelectivityEstimator::compute_ndv_comparison()` is used for integer predicates and immediately calls `Datum::as_double()` on `min`, `max`, and the literal. For `0..=u64::MAX`, `float(u64::MAX - 2) == float(u64::MAX)`, so `d > u64::MAX - 2` is estimated as zero rows even though `u64::MAX` still matches. Please make that fallback integer-safe as well, or bail out to `Unknown` when the bounds/literal are outside exact `f64` precision.

**Comment by sundy-li** [comment]
Blocking issues from review:

1. Skipping synthesized histograms in `scan.rs` is not sufficient for full-range integer stats. Once `derive_stats()` returns `None` for the histogram, `SelectivityEstimator::compute_ndv_comparison()` in `src/query/sql/src/planner/optimizer/ir/stats/selectivity.rs` still converts `min`, `max`, and the literal through `Datum::as_double()`. For `UInt64(0..=u64::MAX)`, values near the upper tail collapse to the same `f64`, so predicates like `d > u64::MAX - 2` are still estimated as zero rows even though `u64::MAX` matches.

2. `has_exact_integer_histogram_bounds()` is off by one. `HistogramBuilder::from_ndv()` models the inclusive integer range by adding `1.0` to `max` after casting to `f64`, so `max == 1 << 53` (and the signed `+/-` boundary) is already unsafe because `max + 1.0 == max` at that point. The guard needs `< 1 << 53` or an equivalent representability check.

File: src/query/sql/src/planner/plans/scan.rs (hunk 2 above)
Comment (sundy-li) [inline src/query/sql/src/planner/plans/scan.rs]: Blocking: skipping synthesized histograms here still leaves full-range integer filters on the lossy `Datum::as_double()` fallback in `SelectivityEstimator::compute_ndv_comparison()`. I reproduced this on the current head with a focused unit test over `ColumnStat { min: 0, max: u64::MAX, ndv: 4, histogram: None }`: `a > 18446744073709551613` estimated `0` rows even though `u64::MAX` still matches. Please make that integer comparison path precision-safe too, or bail out when the bounds/literal exceed exact `f64` integer precision.

File: src/query/sql/src/planner/plans/scan.rs (hunk 3 above)
Comment (sundy-li) [inline src/query/sql/src/planner/plans/scan.rs]: Blocking: this guard is still off by one at the IEEE-754 boundary. `HistogramBuilder::from_ndv()` converts these bounds to `f64` and models the inclusive range with `max + 1.0`; when `max == 1 << 53` (and likewise on the signed `+/-` boundary), `max + 1.0` rounds back to `max`, so the synthesized histogram is still lossy on a range this check currently allows. I reproduced this on the current head with `[2^53 - 2, 2^53]`: `a < 9007199254740991` estimated `2` rows instead of `1`. Please tighten the bound to exclude `1 << 53` or move the inclusive-range math out of float space.

**Comment by sundy-li** [comment]
Review result: fail.

I verified two blocking issues on the current head `ea366936fc` with focused local reproductions:

1. Skipping synthesized histograms for wide integer bounds does not make the fallback comparison path safe. `SelectivityEstimator::compute_ndv_comparison()` still converts `min`, `max`, and the literal through `Datum::as_double()`. With `min = 0`, `max = u64::MAX`, `ndv = 4`, `histogram = None`, the estimate for `a > 18446744073709551613` came back as `0` rows.

2. `has_exact_integer_histogram_bounds()` still admits the `2^53` boundary, but the histogram builder immediately converts the bounds to `f64` and models inclusive integer ranges with `max + 1.0`. At `max == 1 << 53`, that addition rounds back to `max`, so the synthesized histogram is already lossy on an allowed range. With `[2^53 - 2, 2^53]`, the estimate for `a < 9007199254740991` came back as `2` rows instead of `1`.

The inline comments on `scan.rs` point to the exact spots that need changes.

File: src/query/sql/src/planner/plans/scan.rs (hunk 2 above)
Comment (sundy-li) [reply src/query/sql/src/planner/plans/scan.rs]: Fixed in `df88ea4da2`. `has_exact_integer_histogram_bounds()` now excludes the `1 << 53` upper boundary (`max < 1 << 53` for unsigned, signed max also exclusive), so synthesized histograms no longer use the lossy `max + 1.0` path at the IEEE-754 edge. Added `test_derive_stats_skips_histogram_at_f64_precision_boundary` for `[2^53 - 2, 2^53]`.

Comment (sundy-li) [reply src/query/sql/src/planner/plans/scan.rs]: Fixed in `df88ea4da2`. `has_exact_integer_histogram_bounds()` now excludes the `1 << 53` upper boundary (`max < 1 << 53` for unsigned, signed max also exclusive), so synthesized histograms no longer use the lossy `max + 1.0` path at the IEEE-754 edge. Added `test_derive_stats_skips_histogram_at_f64_precision_boundary` for `[2^53 - 2, 2^53]`.

Comment (sundy-li) [reply src/query/sql/src/planner/plans/scan.rs]: Fixed in `df88ea4da2`. `has_exact_integer_histogram_bounds()` now excludes the `1 << 53` upper boundary (`max < 1 << 53` for unsigned, signed max also exclusive), so synthesized histograms no longer use the lossy `max + 1.0` path at the IEEE-754 edge. Added `test_derive_stats_skips_histogram_at_f64_precision_boundary` for `[2^53 - 2, 2^53]`.

File: src/query/sql/src/planner/plans/scan.rs (hunk 3 above)
Comment (sundy-li) [reply src/query/sql/src/planner/plans/scan.rs]: Fixed in `df88ea4da2`. `has_exact_integer_histogram_bounds()` now excludes the `1 << 53` upper boundary (`max < 1 << 53` for unsigned, signed max also exclusive), so synthesized histograms no longer use the lossy `max + 1.0` path at the IEEE-754 edge. Added `test_derive_stats_skips_histogram_at_f64_precision_boundary` for `[2^53 - 2, 2^53]`.

Comment (sundy-li) [reply src/query/sql/src/planner/plans/scan.rs]: Fixed in `df88ea4da2`. `has_exact_integer_histogram_bounds()` now excludes the `1 << 53` upper boundary (`max < 1 << 53` for unsigned, signed max also exclusive), so synthesized histograms no longer use the lossy `max + 1.0` path at the IEEE-754 edge. Added `test_derive_stats_skips_histogram_at_f64_precision_boundary` for `[2^53 - 2, 2^53]`.

File: src/query/sql/src/planner/plans/scan.rs (hunk 2 above)
Comment (sundy-li) [reply src/query/sql/src/planner/plans/scan.rs]: Fixed in `df88ea4da2`. The remaining issue was in `compute_ndv_comparison()`, not just the histogram gate. That fallback now returns `Selectivity::Unknown` whenever the integer min/max range or predicate literal falls outside exact `f64` precision, so full-range `UInt64` stats no longer estimate `a > 18446744073709551613` as zero. Added `test_comparison_with_full_range_uint64_keeps_positive_selectivity` for the reported case.

Comment (sundy-li) [reply src/query/sql/src/planner/plans/scan.rs]: Fixed in `df88ea4da2`. The remaining issue was in `compute_ndv_comparison()`, not just the histogram gate. That fallback now returns `Selectivity::Unknown` whenever the integer min/max range or predicate literal falls outside exact `f64` precision, so full-range `UInt64` stats no longer estimate `a > 18446744073709551613` as zero. Added `test_comparison_with_full_range_uint64_keeps_positive_selectivity` for the reported case.

Comment (sundy-li) [reply src/query/sql/src/planner/plans/scan.rs]: Fixed in `df88ea4da2`. The remaining issue was in `compute_ndv_comparison()`, not just the histogram gate. That fallback now returns `Selectivity::Unknown` whenever the integer min/max range or predicate literal falls outside exact `f64` precision, so full-range `UInt64` stats no longer estimate `a > 18446744073709551613` as zero. Added `test_comparison_with_full_range_uint64_keeps_positive_selectivity` for the reported case.

**Comment by sundy-li** [comment]
Pushed `df88ea4da2` to `sundy-li/fuse-query:fix/issue-19555-planner-u64-overflow`.

Addressed the two remaining blockers from review:
- `Scan::derive_stats()` now excludes the `1 << 53` histogram boundary so synthesized integer histograms do not enter the lossy `max + 1.0` path.
- `SelectivityEstimator::compute_ndv_comparison()` now bails out to `Selectivity::Unknown` when integer bounds or literals exceed exact `f64` precision, so full-range `UInt64` filters no longer estimate zero rows from collapsed tail values.

Added focused regressions for both cases and re-ran the local SQL crate verification listed in the branch summary.

File: src/query/sql/src/planner/optimizer/ir/stats/selectivity.rs [hunk 4]
```diff
@@ -396,6 +421,12 @@ impl SelectivityVisitor<'_> {
         comparison_op: ComparisonOp,
         const_datum: &Datum,
     ) -> Result<Selectivity> {
+        if !has_exact_f64_integer_range(&column_stat.min, &column_stat.max)
+            || !has_exact_f64_integer_value(const_datum)
+        {
+            return Ok(Selectivity::Unknown);
```
Comment (sundy-li) [inline src/query/sql/src/planner/optimizer/ir/stats/selectivity.rs]: This guard is too broad. It also drops comparisons that are still exactly decidable from integer min/max without any unsafe `f64` arithmetic. A narrow large-`UInt64` range reproduces it: with `min = 10000000000000000`, `max = 10000000000000010`, `ndv = 11`, `histogram = None`, `a >= 10000000000000000` should stay at 100% selectivity and `a < 10000000000000000` should stay at 0%, but this path now returns `Unknown` and falls back to the default 20% estimate. We still need the boundary/out-of-range checks here, and only the in-range ratio math needs to bail out.

Comment (sundy-li) [reply src/query/sql/src/planner/optimizer/ir/stats/selectivity.rs]: Fixed in `d649722a9f`. `compute_ndv_comparison()` now classifies integer boundary and out-of-range cases with `Datum::compare()` before the exact-`f64` guard, so large `UInt64` ranges keep exact results for cases like `a >= min` and `a < min`. Only the interior ratio branch still bails out to `Unknown` when the inputs exceed exact `f64` precision. Added `test_comparison_with_large_uint64_range_keeps_exact_boundary_selectivity` for `[10^16, 10^16 + 10]`.

File: src/query/sql/src/planner/optimizer/ir/stats/selectivity.rs [hunk 5]
```diff
@@ -54,6 +54,31 @@ pub const MAX_SELECTIVITY: f64 = 1f64;
...
+        _ => true,
+    }
+}
+
+fn has_exact_f64_integer_range(min: &Datum, max: &Datum) -> bool {
+    match (min, max) {
+        (Datum::UInt(min), Datum::UInt(max)) => {
+            *min < MAX_EXACT_F64_INTEGER && *max < MAX_EXACT_F64_INTEGER
+        }
+        (Datum::Int(min), Datum::Int(max)) => {
```
Comment (sundy-li) [inline src/query/sql/src/planner/optimizer/ir/stats/selectivity.rs:102]: `has_exact_f64_integer_range()` still lets signed ranges like `[-2^53, 0]` through. Both endpoints are individually representable as `f64`, but the span `(max - min + 1)` is `2^53 + 1`, which rounds back to `2^53`. The code below then does `as_double()` math on that inexact span, so a predicate like `a < -1` can still produce `n == 1.0` instead of a value slightly below 1.0. The same boundary check is also copied into `Scan::has_exact_integer_histogram_bounds`, so we would still build float-based histograms for that signed range. Please tighten the guard to validate the span (or switch these paths to integer arithmetic) and add a signed regression test.

**Comment by sundy-li** [comment]
Blocking issue: the new signed exactness guard is still incomplete. Ranges like `[-2^53, 0]` pass the new checks even though `(max - min + 1)` is `2^53 + 1`, which is not exactly representable in `f64`. That means this PR still drives the float-based selectivity and histogram code paths for those signed inputs. Please tighten the guard around representable spans, or move those paths to integer arithmetic, and add a signed regression test.

File: src/query/sql/src/planner/optimizer/ir/stats/selectivity.rs (hunk 5 above)
Comment (sundy-li) [reply src/query/sql/src/planner/optimizer/ir/stats/selectivity.rs:102]: Fixed in `182bd4d05c`. The exact-range guard now validates the inclusive integer span in addition to endpoint precision, so signed ranges like `[-2^53, 0]` no longer enter the float-based selectivity or synthesized-histogram paths. Added regressions for `a < -1` on `[-2^53, 0]` and for skipping synthesized scan histograms on the same signed span.

File: src/query/sql/src/planner/optimizer/ir/stats/selectivity.rs [hunk 6]
```diff
@@ -426,6 +459,15 @@ impl SelectivityVisitor<'_> {
                 return Ok(selectivity);
             }
             (ComparisonOp::LT | ComparisonOp::LTE, _, _) => {
+                if !has_exact_f64_integer_range(&column_stat.min, &column_stat.max)
```
Comment (sundy-li) [inline src/query/sql/src/planner/optimizer/ir/stats/selectivity.rs]: `has_exact_f64_integer_range()` is the right guard for histogram construction, but it is too strict for the NDV fallback below. For example, the range `[2^53-2, 2^53]` still has exact `as_double()` values, so `a < 9007199254740991` can compute an exact `2/3` selectivity with `(value - min + 1) / (max - min + 1)`. This helper rejects that case because `max == MAX_EXACT_F64_INTEGER`, so we now fall back to `Selectivity::Unknown` and the planner uses the default 0.2 instead. Please split the histogram-safe check from the NDV-comparison-safe check, or relax the selectivity guard so exact boundary ranges like this keep their precise estimate.

**Comment by sundy-li** [comment]
Blocking issue:

The new exact-range guard is too strict at the `2^53` boundary. In `src/query/sql/src/planner/optimizer/ir/stats/selectivity.rs:78-90` you reject `UInt64` ranges unless both endpoints are `< 1 << 53` (and `Int64` ranges unless `max < 1 << 53`), but integers up to and including `2^53` are still exactly representable in `f64`.

That turns exact ranges into `Selectivity::Unknown` in `compute_ndv_comparison` and also suppresses histogram construction in `src/query/sql/src/planner/plans/scan.rs:309-324`. A concrete repro is `min = 2^53 - 2`, `max = 2^53`, `ndv = 3`, predicate `a < 9007199254740991`: the exact answer is `2/3`, but the new code falls back to the default 0.2 selectivity (18 / 90 rows in a local repro).

Please make the boundary check inclusive (`<= 1 << 53` / `..=limit`) and add a regression test for an interior predicate on a range ending at `2^53`, otherwise this change regresses planner estimates on safe integer ranges.

File: src/query/sql/src/planner/optimizer/ir/stats/selectivity.rs (hunk 6 above)
Comment (sundy-li) [inline src/query/sql/src/planner/optimizer/ir/stats/selectivity.rs]: Blocking: this shared exact-range guard is too strict for the NDV comparison path. It rejects safe `UInt64` ranges that end exactly at `1 << 53`, even though those values are still exactly representable in `f64` and the arithmetic below stays precise when the span is small.

Concrete repro on the current head `182bd4d05c`: `min = 9007199254740990`, `max = 9007199254740992`, `ndv = 3`, `histogram = None`, predicate `a < 9007199254740991`. The exact selectivity is `1/3` (30 of 90 rows), but `has_exact_f64_integer_range()` returns false here, so we bail out to `Selectivity::Unknown` and end up with the default `0.2` estimate (18 rows).

I verified that with a focused local unit test against this head. Histogram synthesis does need a stricter boundary because it does `max + 1.0`, but the NDV fallback should not reuse that same `< 1 << 53` cutoff. Please split the guards or otherwise allow the exact `2^53` boundary when the comparison math remains representable.

**Comment by sundy-li** [comment]
Pushed `aec57c33ec` to `sundy-li/fuse-query:fix/issue-19555-planner-u64-overflow`.

Addressed the remaining exact-boundary review feedback:
- split the integer exactness guard so synthesized histograms still require the stricter `< 1 << 53` bound, while NDV comparison math accepts exact small spans ending at `1 << 53`
- added `test_comparison_with_uint64_range_ending_at_f64_boundary_keeps_selectivity` for the reported `[2^53 - 2, 2^53]` repro

Local `cargo fmt --all --check` passed. Focused `cargo test -p databend-common-sql ...` is blocked in this container because the configured Linux toolchain expects `clang`/`ld`, which are not installed here; the new regression is included for CI coverage.

**Comment by sundy-li** [comment]
Blocking issue: this changes planner statistics behavior but only adds Rust unit/integration coverage. Per the repository testing guidance for planner changes, please add a deterministic SQL regression (with expected output) that exercises the failing path end to end. The new tests in `src/query/sql/tests/it/planner/fixture.rs` are useful, but they do not replace the planner-level SQL regression coverage expected for a submit-ready fix.
